pkg/registry/pod/rest.go

```diff
@@ -81,9 +81,28 @@ func (podStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) fiel
 	return append(errorList, validation.ValidatePodUpdate(obj.(*api.Pod), old.(*api.Pod))...)
 }
 
-// CheckGracefulDelete allows a pod to be gracefully deleted.
+// CheckGracefulDelete allows a pod to be gracefully deleted. It updates the DeleteOptions to
+// reflect the desired grace value.
 func (podStrategy) CheckGracefulDelete(obj runtime.Object, options *api.DeleteOptions) bool {
-	return false
+	if options == nil {
+		return false
+	}
+	pod := obj.(*api.Pod)
+	// use the user provided value
+	if options.GracePeriodSeconds != nil {
+		period := int64(*options.GracePeriodSeconds)
+		// if the pod is not currently being deleted, or the value isn't set, or the user value is less than the stored value
+		if pod.Spec.TerminationGracePeriodSeconds == nil || pod.DeletionTimestamp == nil || period < *pod.Spec.TerminationGracePeriodSeconds {
+			pod.Spec.TerminationGracePeriodSeconds = &period
```

Comment:
Ok, I can see the point that the DeleteOptions are modifying the desired state, but it definitely adds a wrinkle to configuration reconciliation #1702.

This may not be an issue for controller-generated pods, since they'll manage terminationGracePeriodSeconds in the pod template rather than in the pods themselves, but I want to get the principle sorted out.

This isn't the only known instance where modifying an initialized field will happen -- e.g., auto-scaling will modify an initial replicas count and auto-sizing will modify initial cpu and memory values -- but either those values would be defaults from the beginning, or initial value "hints" could be provided via an annotation or in the auto-scaling/sizing configs. In either case, the config mechanism would know that the user didn't intend to configure those fields, since the user's config would not contain the initial values.

If a user did have a pod managed by a declarative config file, they might legitimately want to declaratively update terminationGracePeriodSeconds, until it is set imperatively via DeleteOptions. I can tell you from experience that there will be scenarios where someone wants to update a pod during a (sufficiently long) grace period. 

The (declarative) semantics that I'm accustomed to, and that proposed in #1702, #1007, and elsewhere, is that removal of a field from a config implies that it should be restored to its default value before dropping it from the set of declaratively managed fields. If it were acceptable to require users to explicitly specify the updated value they expected or to explicitly nullify the field, as in a [patch](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/api-conventions.md#strategic-merge-patch), in at least one declarative update, then that would make patch generation significantly simpler. We could potentially treat omission of a field simply as dropping it from the set of declaratively managed fields, though I have no idea whether users would find that unintuitive. We'd still need to detect declarative removal of map and list elements, IMO, such as removal of a label or of a container or volume, so we couldn't just use the annotation containing the config to determine that the configuration changed, as I suggested for Kubelet's mirror pods (#8683). 

One challenge with using a magic map key as a  json/yaml annotation in user configs to indicate that certain fields shouldn't be diff'ed during patch generation (e.g., `$noupdate`) is that we can only use the same key once per map. I suppose it could specify a list of field names.

cc @ghodss for his thoughts